pkg/store/consul/transaction/transaction.go

```diff
@@ -11,90 +12,86 @@ import (
 	ErrAlreadyCommitted  = errors.New("this transaction has already been committed")
+	ErrNoTransaction     = errors.New("no transaction was opened in the passed Context")
+
+	contextKey = contextKeyType{}
 )
 
 type Tx struct {
-	kvOps            *api.KVTxnOps
-	alreadyCommitted bool
+	kvOps *api.KVTxnOps
 
 	commitHooks []func()
 }
 
-func New() *Tx {
-	return &Tx{
+func New(ctx context.Context) (context.Context, context.CancelFunc) {
+	ctx, cancelFunc := context.WithCancel(ctx)
+	ctx = context.WithValue(ctx, contextKey, &Tx{
 		kvOps: new(api.KVTxnOps),
-	}
+	})
+	return ctx, cancelFunc
 }
 
-func (c *Tx) Add(op api.KVTxnOp) error {
-	if c.alreadyCommitted {
-		return ErrAlreadyCommitted
+func Add(ctx context.Context, op api.KVTxnOp) error {
+	txn, err := getTxnFromContext(ctx)
+	if err != nil {
+		return err
 	}
 
-	if len(*c.kvOps) == maxAllowedOperations {
+	if len(*txn.kvOps) == maxAllowedOperations {
 		return ErrTooManyOperations
 	}
-	*c.kvOps = append(*c.kvOps, &op)
+	*txn.kvOps = append(*txn.kvOps, &op)
 
 	return nil
 }
 
 // AddCommitHook adds a function that should be run when the transaction is
 // committed.  This is useful for cleaning up resources, e.g. consul sessions
 // that had to be opened as part of transaction setup.
-func (c *Tx) AddCommitHook(f func()) error {
-	if c.alreadyCommitted {
-		return ErrAlreadyCommitted
-	}
-
-	c.commitHooks = append(c.commitHooks, f)
-	return nil
-}
-
-func (c *Tx) Append(newTxn *Tx) error {
-	if c.alreadyCommitted {
-		return ErrAlreadyCommitted
-	}
-
-	if newTxn.alreadyCommitted {
-		return ErrAlreadyCommitted
-	}
-
-	if len(*c.kvOps)+len(*newTxn.kvOps) > 64 {
-		return ErrTooManyOperations
-	}
-
-	for _, op := range *newTxn.kvOps {
-		*c.kvOps = append(*c.kvOps, op)
+func AddCommitHook(ctx context.Context, f func()) error {
+	txn, err := getTxnFromContext(ctx)
+	if err != nil {
+		return err
 	}
 
+	txn.commitHooks = append(txn.commitHooks, f)
 	return nil
 }
 
 type Txner interface {
 	Txn(txn api.KVTxnOps, q *api.QueryOptions) (bool, *api.KVTxnResponse, *api.QueryMeta, error)
 }
 
-func (c *Tx) Commit(txner Txner) error {
-	if c.alreadyCommitted {
-		return ErrAlreadyCommitted
+// Commit attempts to run all of the kv operations in the context's
+// transaction. The cancel function with which the context was created must
+// also be passed to guarantee that the transaction won't be applied twice
+func Commit(ctx context.Context, cancel context.CancelFunc, txner Txner) error {
+	txn, err := getTxnFromContext(ctx)
+	if err != nil {
+		return nil
 	}
 
-	ok, resp, _, err := txner.Txn(*c.kvOps, nil)
+	ok, resp, _, err := txner.Txn(*txn.kvOps, nil)
 	if err != nil {
 		return util.Errorf("transaction failed: %s", err)
 	}
 
-	c.runCommitHooks()
-	// Set this after we know err == nil because otherwise it could have
+	txn.runCommitHooks()
+
+	// Run this after we know err == nil because otherwise it could have
 	// been a retry-able TCP error for example
-	c.alreadyCommitted = true
+	cancel()
```

Comment:
It's the opposite, i'm using `cancel()` here to mark the transaction as completed so it can't be used again. This is enforced by checking `<-ctx.Done()` in `getTxnFromContext()`.

I don't want to cancel if there was an error, because the caller might want to retry the error (e.g. if consul temporarily unavailable, don't want to go through the work of rebuilding the whole transaction)